ambry-router/src/main/java/com.github.ambry.router/DefaultGCMCryptoService.java

```diff
@@ -0,0 +1,298 @@
+   *
+   *  content    - The actual IV
+   *
+   */
+  private static class IVRecord_Format_V1 {
+    private static final int IV_SIZE_FIELD_IN_BYTES = Integer.BYTES;
+    private static Logger logger = LoggerFactory.getLogger(IVRecord_Format_V1.class);
+
+    /**
+     * Returns the IV record size for the given iv
+     * @param iv the iv for which IV record size is requested
+     * @return the size of the IV record size for the given iv
+     */
+    private static int getIVRecordSize(byte[] iv) {
+      return Version_Field_Size_In_Bytes + IV_SIZE_FIELD_IN_BYTES + iv.length;
+    }
+
+    /**
+     * Serialize IV into {@link ByteBuffer}
+     * @param outputBuffer the {@link ByteBuffer} to which the IV needs to be serialized
+     * @param iv the iv of type byte array that needs to be serialized
+     */
+    private static void serializeIVRecord(ByteBuffer outputBuffer, byte[] iv) {
+      outputBuffer.putShort(IVRecord_Version_V1);
+      outputBuffer.putInt(iv.length);
+      outputBuffer.put(iv);
+    }
+
+    /**
+     * Deserializes the iv from the stream
+     * @param dataStream the {@link DataInputStream} from which iv needs to be deserialized
+     * @return the iv thus deserialized from the stream
+     * @throws IOException
+     * @throws MessageFormatException
+     */
+    private static byte[] deserializeIVRecord(DataInputStream dataStream) throws IOException, MessageFormatException {
+      int ivSize = dataStream.readInt();
+      return Utils.readBytesFromStream(dataStream, ivSize);
+    }
+  }
+
+  /**
+   *  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+   * |         |           |            |               |              |
+   * | version | key size  |  content   |  key gen algo | key gen Algo |
+   * |(2 bytes)| (4 bytes) |  (n bytes) |      size     |   (n bytes)  |
+   * |         |           |            |   (4 bytes)   |              |
+   *  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+   *  version             - The version of the Key Record
+   *
+   *  key size            - The size of the key
+   *
+   *  content             - The actual key
+   *
+   *  key gen algo size   - Size of the key gen algo
+   *
+   *  key gen algo        - The actual key gen algo
+   */
+  private static class KeyRecord_Format_V1 {
+    private static final int KEY_SIZE_FIELD_IN_BYTES = Integer.BYTES;
+    private static final int KEY_GEN_ALGO_SIZE_FIELD_IN_BYTES = Integer.BYTES;
+
+    /**
+     * Returns the key record size for the given key and key gen algo
+     * @param key the key for which key record size is requested
+     * @param keyGenAlgo the key gen algo used to generate the key
+     * @return the size of key record for the given key and key gen algo
+     */
+    private static int getKeyRecordSize(byte[] key, String keyGenAlgo) {
+      return Version_Field_Size_In_Bytes + KEY_SIZE_FIELD_IN_BYTES + key.length + KEY_GEN_ALGO_SIZE_FIELD_IN_BYTES
```

Comment:
There is a function `getIntStringLength()` in `Utils` that can return the size of serialized `String`.